File: solutions/python/react/1/react.py
class InputCell:
    def __init__(self, value):
        self._value = value
        self._observers = []

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, value):
        if self._value != value:
            self._value = value
            # Topological sort of all dependent cells
            observers = self._get_all_observers()
            
            # Keep track of last reported values to ensure change is detected
            # Actually, ComputeCell already stores last_reported_value.
            # We just need to trigger the update and check.
            
            # The key is to propagate updates only IF the value actually changes
            # This is complex because we need to know if the value changed.
            # Let's use a two-pass approach: 1. Update, 2. Notify.
            
            for observer in observers:
                observer.update()
                
            # Trigger callbacks for those that changed
            for observer in observers:
                observer._check_callbacks()

    def _get_all_observers(self):
        observers = {}
        def collect(cell, depth):
            for observer in cell._observers:
                if observer not in observers or observers[observer] < depth:
                    observers[observer] = depth
                    collect(observer, depth + 1)
        collect(self, 1)
        return sorted(observers.keys(), key=lambda x: observers[x])


class ComputeCell:
    def __init__(self, inputs, compute_function):
        self._inputs = inputs
        self._compute_function = compute_function
        self._value = self._compute_function([i.value for i in self._inputs])
        self._callbacks = []
        self._observers = []
        self._last_reported_value = self._value
        
        for input_cell in inputs:
            input_cell._observers.append(self)

    def update(self):
        self._value = self._compute_function([i.value for i in self._inputs])
        
    def _check_callbacks(self):
        # We need to only trigger if the value actually changed.
        # But we also need to store the value as it was before this update, 
        # to compare correctly.
        if self._value != self._last_reported_value:
            for callback in list(self._callbacks):
                if callback in self._callbacks:
                    callback(self._value)
            self._last_reported_value = self._value

    @property
    def value(self):
        return self._value

    def add_callback(self, callback):
        self._callbacks.append(callback)

    def remove_callback(self, callback):
        if callback in self._callbacks:
            self._callbacks.remove(callback)
    

File: solutions/javascript/react/1/react.js
export class InputCell {
  constructor(value) {
    this._value = value;
    this._dependents = [];
  }

  get value() {
    return this._value;
  }

  setValue(value) {
    if (this._value !== value) {
      this._value = value;
      const affected = [];
      const visited = new Set();
      this._collect(affected, visited);
      for (let i = affected.length - 1; i >= 0; i--) {
        affected[i]._recompute();
      }
    }
  }

  _collect(affected, visited) {
    for (const dep of this._dependents) {
      if (!visited.has(dep)) {
        visited.add(dep);
        dep._collect(affected, visited);
        affected.push(dep);
      }
    }
  }

  _addDependent(cell) {
    this._dependents.push(cell);
  }
}

export class ComputeCell {
  constructor(inputCells, fn) {
    this._inputs = inputCells;
    this._fn = fn;
    this._callbacks = [];
    this._dependents = [];
    this._value = this._compute();
    for (const cell of inputCells) {
      cell._addDependent(this);
    }
  }

  get value() {
    return this._value;
  }

  _compute() {
    return this._fn(this._inputs);
  }

  _recompute() {
    const newValue = this._compute();
    if (this._value !== newValue) {
      this._value = newValue;
      for (const cb of this._callbacks) {
        cb._fire(this);
      }
    }
  }

  _collect(affected, visited) {
    for (const dep of this._dependents) {
      if (!visited.has(dep)) {
        visited.add(dep);
        dep._collect(affected, visited);
        affected.push(dep);
      }
    }
  }

  _addDependent(cell) {
    this._dependents.push(cell);
  }

  addCallback(cb) {
    this._callbacks.push(cb);
  }

  removeCallback(cb) {
    const idx = this._callbacks.indexOf(cb);
    if (idx !== -1) {
      this._callbacks.splice(idx, 1);
    }
  }
}

export class CallbackCell {
  constructor(fn) {
    this._fn = fn;
    this._values = [];
  }

  get values() {
    return this._values;
  }

  _fire(cell) {
    this._values.push(this._fn(cell));
  }
}
